operator/controllers/btpoperator_controller.go

```diff
@@ -95,25 +103,236 @@ var (
 	labelFilter = client.MatchingLabels{labelKeyForChart: operatorName}
 )
 
-type btpOperatorGvk struct {
-	APIVersion string `yaml:"apiVersion"`
-	Kind       string `yaml:"kind"`
-}
-
 // BtpOperatorReconciler reconciles a BtpOperator object
 type BtpOperatorReconciler struct {
 	client.Client
 	*rest.Config
 	Scheme                *runtime.Scheme
+	currentVersion        string
+	wereUpdateCheckDone   bool
 	WaitForChartReadiness bool
 }
 
+func (r *BtpOperatorReconciler) handleUpdate(ctx context.Context, cr *v1alpha1.BtpOperator) error {
+	logger := log.FromContext(ctx)
+
+	configMap := r.buildBtpManagerConfigMap()
+	err := r.Get(ctx, client.ObjectKeyFromObject(configMap), configMap)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	} else if err != nil && k8serrors.IsNotFound(err) {
+		configMap = nil
+	}
+
+	versionChanged, err := r.storeChartDetails(ctx, configMap)
+	if err != nil {
+		return err
+	}
+
+	logger.Info(fmt.Sprintf("version changed? -> %t", versionChanged))
+
+	if versionChanged {
+		ok, err := r.HandleReadyState(ctx, cr)
+		if !ok {
+			return fmt.Errorf("consistency check failed, deletion of orphanes resoures will be skipped")
+		}
+		if err != nil {
+			return err
+		}
+
+		if err := r.deleteOrphanedResources(ctx, configMap); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) createDefaultNamespaceIfNeeded(ctx context.Context) error {
+	namespace := &corev1.Namespace{}
+	namespace.Name = ChartNamespace
+	err := r.Get(ctx, client.ObjectKeyFromObject(namespace), namespace)
+	if k8serrors.IsNotFound(err) {
+		err = r.Create(ctx, namespace)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) buildBtpManagerConfigMap() *corev1.ConfigMap {
+	configMap := &corev1.ConfigMap{}
+	configMap.Namespace = ChartNamespace
+	configMap.Name = btpManagerConfigMap
+	return configMap
+}
+
+func (r *BtpOperatorReconciler) storeChartDetails(ctx context.Context, configMap *corev1.ConfigMap) (bool, error) {
+	newChartVersion, err := ymlutils.ExtractStringValueFromYamlForGivenKey(fmt.Sprintf("%s/Chart.yaml", ChartPath), "version")
+	if err != nil {
+		return false, err
+	}
+
+	newGvks, err := ymlutils.GatherChartGvks(ChartPath)
+	if err != nil {
+		return false, err
+	}
+
+	err = r.createDefaultNamespaceIfNeeded(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	if configMap == nil {
+		configMap = r.buildBtpManagerConfigMap()
```

Comment:
nit: `buildBtpManagerConfigMap` is called before in `handleUpdate` L119, we can just check if `Data` fields is empty